File: backend/app/api/v1/endpoints/audit.py
"""Audit log endpoints for administrators."""

import math

from fastapi import APIRouter, Depends, Query

from app.dependencies import DBSession, require_role
from app.schemas.audit import AuditLogListResponse, AuditLogResponse
from app.services.audit import list_audit_logs

router = APIRouter()


@router.get(
    "",
    response_model=AuditLogListResponse,
    summary="List audit logs",
    description="List audit logs with optional filters. Admin only.",
)
async def get_audit_logs(
    db: DBSession,
    _current_user=Depends(require_role(["administrator"])),
    action: str | None = Query(None, description="Filter by action type"),
    user_id: str | None = Query(None, description="Filter by user ID"),
    entity_type: str | None = Query(None, description="Filter by entity type"),
    date_from: str | None = Query(None, description="Filter from date (ISO 8601)"),
    date_to: str | None = Query(None, description="Filter to date (ISO 8601)"),
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
):
    """List audit logs with filtering and pagination."""
    logs, total = await list_audit_logs(
        db,
        action=action,
        user_id=user_id,
        entity_type=entity_type,
        date_from=date_from,
        date_to=date_to,
        page=page,
        limit=limit,
    )
    return AuditLogListResponse(
        items=[
            AuditLogResponse(
                id=log.id,
                timestamp=log.timestamp,
                user_id=log.user_id,
                user_email=log.user_email,
                role=log.role,
                ip_address=log.ip_address,
                action=log.action,
                entity_type=log.entity_type,
                entity_id=log.entity_id,
                detail=log.detail,
                result=log.result,
            )
            for log in logs
        ],
        total=total,
        page=page,
        limit=limit,
        pages=math.ceil(total / limit) if total > 0 else 0,
    )

File: frontend/src/api/audit.ts
import { apiClient } from './client';

export interface AuditLogEntry {
  id: number;
  timestamp: string;
  userId: string | null;
  userEmail: string | null;
  role: string | null;
  ipAddress: string | null;
  action: string;
  entityType: string | null;
  entityId: string | null;
  detail: string | null;
  result: string;
}

export interface AuditLogListResponse {
  items: AuditLogEntry[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface AuditLogFilters {
  action?: string;
  userId?: string;
  entityType?: string;
  dateFrom?: string;
  dateTo?: string;
  page?: number;
  limit?: number;
}

export const auditApi = {
  list: async (filters: AuditLogFilters = {}): Promise<AuditLogListResponse> => {
    const params = new URLSearchParams();
    if (filters.action) params.append('action', filters.action);
    if (filters.userId) params.append('user_id', filters.userId);
    if (filters.entityType) params.append('entity_type', filters.entityType);
    if (filters.dateFrom) params.append('date_from', filters.dateFrom);
    if (filters.dateTo) params.append('date_to', filters.dateTo);
    if (filters.page) params.append('page', String(filters.page));
    if (filters.limit) params.append('limit', String(filters.limit));

    const response = await apiClient.get<AuditLogListResponse>(
      `/v1/audit-logs?${params.toString()}`
    );
    return response.data;
  },
};
